hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java

```diff
@@ -429,25 +428,21 @@ public void mergeArchiveFiles(List<FileStatus> compactCandidate) throws IOExcept
         .collect(Collectors.groupingBy(i -> Pair.of(i.getTimestamp(),
             HoodieInstant.getComparableAction(i.getAction()))));
 
-    // If metadata table is enabled, do not archive instants which are more recent than the last compaction on the
-    // metadata table.
-    if (config.isMetadataTableEnabled()) {
-      try (HoodieTableMetadata tableMetadata = HoodieTableMetadata.create(table.getContext(), config.getMetadataConfig(),
-          config.getBasePath(), FileSystemViewStorageConfig.SPILLABLE_DIR.defaultValue())) {
-        Option<String> latestCompactionTime = tableMetadata.getLatestCompactionTime();
```

Comment:
Hello @nsivabalan , i see that you put the code here, is there any special reason that the data set timeline archival should be in front of the metadata latest compaction time ?

Reply:
yes. 
CC @prashantwason 
We have documented the reasoning here https://issues.apache.org/jira/browse/HUDI-2458

Reply:
With the current metadata table design, we validate the deltacommits to read for metadata table based on which completed commits are present in the dataset. So dataset should never be archiving the instants before compaction on metadata table.

Reply:
Thanks, I read the code and this patch does not break the original limitations. Because the metadata table commits before dataset, with the new patch, the instants on the dataset timeline must also be in the metadata timeline, so we can still do the check for dataset instant existence.

Reply:
Hello, @nsivabalan  can you help confirm this ?

Reply:
There can be the case where deltacommit on the metadata table succeeds but commit on dataset fails (inflight -> complete transition fails). 

In this case, we should not be using the deltacommit on the metadata table and to check that we need commits on the dataset to not be archived. If you archive the commits on the datasets indeendantly then we cannot differentiate between the error case above and the case where commit is missing from dataset timeline because it was archived.

Reply:
sorry. is this addressed? if not, can you help me understand how we will overcome the above scenario if we remove this code block ? 

bcoz, our filtering of additional commits in metadata table happens with LogRecordReader. which is effective only w/ latest file slice. Once compaction kicks in, we may never know what all commits have gone into the compacted base file.

Reply:
> There can be the case where deltacommit on the metadata table succeeds but commit on dataset fails (inflight -> complete transition fails).

When the job restarts, the last commit would be rollbacked first right ? And the view of the metadata table would be fixed then, so this should not be a problem.

In any case, the metadata table was read with latest filesystem view which includes the delta commits logs files, i don't  know why we address metadata compaction here because compaction does not affect the table records.

Reply:
>> When the job restarts, the last commit would be rollbacked first right ?
The last commit failed on the dataset but succeed on the metadata table. So yes it will be rolled back on the dataset eventually - depends on the settings  (EAGER vs LAZY rollbacks). 

Also we need to support the readers - they need to ignore the deltacommit. There can be a delay between the failed job and the retry and readers should read consistent data during that time.

Reply:
> Also we need to support the readers - they need to ignore the deltacommit

That is not the case for current metadata table reader i guess, and in personal i think this restriction is too much limited, doesn't the fail rollback already fixed the metadata table then after the rollback metadata was synced ?

Reply:
> > There can be the case where deltacommit on the metadata table succeeds but commit on dataset fails (inflight -> complete transition fails).
> 
> When the job restarts, the last commit would be rollbacked first right ? And the view of the metadata table would be fixed then, so this should not be a problem.
> 

Last commit may not be rolled back immediately for the scenarios of single writer with async table services and multi-writer, since LAZY should be set for `hoodie.cleaner.policy.failed.writes`.

Reply:
Sorry i don't understand why the cleaner policy can affect this, if the rollback metadata was synced, the metadata file list can be trusted right ? And the corrupt files should then be ignored automatically.